### PR TITLE
fix: static cast uniqueDepth to int to avoid ull underflow

### PR DESCRIPTION
**Pull request by ahuber21**
## Description

Remove some inconsistent assignments and comparisons between signed and unsigned (long) ints by explicitly downcasting to int and adding a boundary check. All if this just addresses theoretical over- and underflow scenarios because in reality the SHAP algos never call these internal functions with such exotic values.
Correctness is confirmed by the SHAP value checks in sklearnex.

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

**Comment by ahuber21** [comment]
We don't have specific unit tests for SHAP value correctness in oneDAL (only in sklearnex) and it would be very difficult to add it unfortunately. I checked the correctness manually though.

File: cpp/daal/src/algorithms/dtrees/gbt/treeshap.cpp [hunk 1]
```diff
@@ -34,44 +34,60 @@ namespace v0
...
+
+    uniquePath[depth].featureIndex  = featureIndex;
+    uniquePath[depth].zeroFraction  = zeroFraction;
+    uniquePath[depth].oneFraction   = oneFraction;
+    uniquePath[depth].partialWeight = (depth == 0 ? 1.0f : 0.0f);
 
-    const float constant = 1.0f / static_cast<float>(uniqueDepth + 1);
-    for (int i = uniqueDepth - 1; i >= 0; --i)
+    const float constant = 1.0f / static_cast<float>(depth + 1);
+    for (int i = depth - 1; i >= 0; --i)
```
Comment (Vika-F) [inline cpp/daal/src/algorithms/dtrees/gbt/treeshap.cpp:51]: Coverity report says there are problems when `uniqueDepth == 0` due to that `int i = uniqueDepth - 1` initialization.

It would be less invasive just to add a check for (uniqueDepth > 0) here I think.

Comment (avolkov-intel) [reply cpp/daal/src/algorithms/dtrees/gbt/treeshap.cpp:51]: I think we have different hits for both, so I like this solution with static_cast more

**Review by Vika-F — APPROVED** [review]

**Review by icfaust — APPROVED** [review]
Thank you for this + the comments in the code.
